New FWD tracking mode based on FST-track finding

jdbrice opened the pull request:
These updates provide a new FWD tracking mode in which the track finding is one with FST information.
Tracks are then fit to valid Seeds and then (optionally) any FTT hits are added to the track if found along the trajectory. 

This mode is needed for any analysis of real data at this point. That would include fast offline for ongoing data taking.

jdbrice commented:
Note, this PR includes the changes from PR #569 but the only changes relevant here are those in StRoot/StFwdTrackMaker and StRoot/StFwdUtils/

jdbrice commented:
Also the CI jobs seem to be failing on code unrelated to my changes (things in StEvent)

plexoos commented:
These errors should be fixed first:
```
2023-07-25T19:47:18.5349686Z #10 1453.7 g++ -m64 -fPIC -pipe -Wall -Woverloaded-virtual -std=c++0x -Wno-long-long -pthread -Wno-deprecated-declarations -Werror -O2 -g -falign-loops -falign-jumps -falign-functions -Dsl79_gcc485 -D__ROOT__ -DNEW_DAQ_READER -I. -IStRoot -I.sl79_gcc485/include -I/opt/software/linux-scientific7-x86_64/gcc-4.8.5/root-5.34.38-l3v6vso6qgojm4l2ctwjojs6trbt4hpn/include -c .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx -o .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.o
2023-07-25T19:47:19.4377677Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx: In member function 'void StFwdAnalysisMaker::ProcessFwdTracks()':
2023-07-25T19:47:19.4378676Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:191:54: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4379329Z #10 1454.6          for( int i = 0; i < fcs->clusters(iDet).size(); i++){
2023-07-25T19:47:19.4379659Z #10 1454.6                                                       ^
2023-07-25T19:47:19.4380367Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:245:68: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4381434Z #10 1454.6          for ( int iEcal = 0; iEcal < fwdTrack->ecalClusters().size(); iEcal++ ){
2023-07-25T19:47:19.4381970Z #10 1454.6                                                                     ^
2023-07-25T19:47:19.4382676Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:263:62: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4383319Z #10 1454.6                  for( int i = 0; i < fcs->clusters(iDet).size(); i++){
2023-07-25T19:47:19.4383665Z #10 1454.6                                                               ^
2023-07-25T19:47:19.4384345Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:301:62: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4384994Z #10 1454.6                  for( int i = 0; i < fcs->clusters(iDet).size(); i++){
2023-07-25T19:47:19.4385329Z #10 1454.6                                                               ^
2023-07-25T19:47:19.4385933Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:235:15: error: unused variable 'c' [-Werror=unused-variable]
2023-07-25T19:47:19.4386330Z #10 1454.6          float c[9];
2023-07-25T19:47:19.4386574Z #10 1454.6                ^
2023-07-25T19:47:19.4387294Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx: In member function 'void StFwdAnalysisMaker::ProcessFwdMuTracks()':
2023-07-25T19:47:19.4388295Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:413:55: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4388939Z #10 1454.6              for( int i = 0; i < fcs->numberOfClusters(); i++){
2023-07-25T19:47:19.4389281Z #10 1454.6                                                        ^
2023-07-25T19:47:19.4389984Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:437:70: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4390652Z #10 1454.6          for ( size_t i = 0; i < muFwdTrack->mEcalClusters.GetEntries(); i++ ){
2023-07-25T19:47:19.4391042Z #10 1454.6                                                                       ^
2023-07-25T19:47:19.4391851Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:455:70: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
2023-07-25T19:47:19.4392547Z #10 1454.6          for ( size_t i = 0; i < muFwdTrack->mHcalClusters.GetEntries(); i++ ){
2023-07-25T19:47:19.4392918Z #10 1454.6                                                                       ^
2023-07-25T19:47:19.4393526Z #10 1454.6 .sl79_gcc485/OBJ/StRoot/StFwdUtils/StFwdAnalysisMaker.cxx:402:15: error: unused variable 'c' [-Werror=unused-variable]
2023-07-25T19:47:19.4394034Z #10 1454.6          float c[9];
2023-07-25T19:47:19.4394264Z #10 1454.6                ^
```

klendathu2k commented on StRoot/StEvent/StFttCluster.h:
```diff
@@ -48,17 +61,23 @@ class StFttCluster : public StObject {
+    Int_t   mIndexMaxStrip = -999;// index of the strip with maximum ADC
+    Float_t mMaxStripCenter = -999;// Strip center of strip with maximum ADC
+    Float_t mMaxStripLeftEdge = -999;//left edge of the strip with maximum ADC
+    Float_t mMaxStripRightEdge = -999;//right edge of the strip with maximum ADC
```
Line 85, increase ClassDef version to 3.

jdbrice replied:
See comment below, all changes related to FTT reverted since they should be handled in PR #569 .
Done with https://github.com/star-bnl/star-sw/pull/571/commits/6088f974a30af0141f460b3fdd3183ae4f7c66ec

klendathu2k commented on StRoot/StEvent/StFttPoint.h:
```diff
@@ -36,17 +38,21 @@ class StFttPoint : public StObject {
     Float_t  mX=0.0;         // x-position in local coordinate
     Float_t  mY=0.0;         // y-position in local coordinate
+    Float_t mD1=0.0;
+    Float_t mD2=0.0;
```
Class layout has changed, increase ClassDef to version 2.

jdbrice replied:
See comment below, all changes related to FTT reverted since they should be handled in PR #569 .
Done with https://github.com/star-bnl/star-sw/pull/571/commits/6088f974a30af0141f460b3fdd3183ae4f7c66ec

klendathu2k commented on StRoot/StEvent/StFwdTrack.h:
```diff
@@ -167,11 +173,11 @@ class StFwdTrack : public StObject {
     float mPval;
     short mCharge;
-    StThreeVectorD mPrimaryMomentum;
+    StThreeVectorD mPrimaryMomentum; 
```
Memory layout changes, ClassDef should be version 3.

jdbrice replied:
See comment below, all changes related to FTT reverted since they should be handled in PR #569 .
Done with https://github.com/star-bnl/star-sw/pull/571/commits/6088f974a30af0141f460b3fdd3183ae4f7c66ec

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -38,7 +38,7 @@ StFttClusterMaker::StFttClusterMaker( const char* name )
   mFttDb( nullptr )
 {
-    LOG_DEBUG << "StFttClusterMaker::ctor"  << endm;
+    LOG_INFO << "StFttClusterMaker::ctor"  << endm;
```
Is LOG_INFO really appropriate here?

jdbrice replied:
See comment below, all changes related to FTT reverted since they should be handled in PR #569 .
Done with https://github.com/star-bnl/star-sw/pull/571/commits/6088f974a30af0141f460b3fdd3183ae4f7c66ec

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.h:
```diff
@@ -24,11 +24,11 @@ class StFttClusterMaker: public StMaker {
-    int  FinishRun( int );
-    int  Finish();
-    int  Make();
+    Int_t  Init();
```
Basic c-types (int, float, double, etc...) are preferred in makers.  Int_t, Float_t, etc... should be used in classes which are made persistent.

jdbrice replied:
See comment below, all changes related to FTT reverted since they should be handled in PR #569 .
Done with https://github.com/star-bnl/star-sw/pull/571/commits/6088f974a30af0141f460b3fdd3183ae4f7c66ec

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.h:
```diff
@@ -47,12 +47,12 @@ class StFttClusterMaker: public StMaker {
 
 
-    ClassDef( StFttClusterMaker, 1 )
+    ClassDef( StFttClusterMaker, 2 )
```
I do not see a change to the memory layout here.  So ClassDef does not need to be incremented.  And since the maker will not be streamed, you can use ClassDef( StFttClusterMaker, 0).

jdbrice replied:
See comment below, all changes related to FTT reverted since they should be handled in PR #569 .
Done with https://github.com/star-bnl/star-sw/pull/571/commits/6088f974a30af0141f460b3fdd3183ae4f7c66ec

klendathu2k reviewed (commented):
Few initial comments made.  Will review in more detail tomorrow.

jdbrice commented:
@klendathu2k I will remove everything StFtt related from this PR. Everything you suggest is valid - but it I've addressed it in PR #569 instead of here

jdbrice commented:
Well, I thought that PR #538 (Add Fwd Tracks to MuDst) was already merged. It was submitted in May but has not received any review. So I removed MuDST from the PR also (done in https://github.com/star-bnl/star-sw/pull/571/commits/ae1dae2f96d009378c6be9dcf21e19c0f523380d)

plexoos commented:
> Well, I thought that PR #538 (Add Fwd Tracks to MuDst) was already merged. It was submitted in May but has not received any review. So I removed MuDST from the PR also (done in [ae1dae2](https://github.com/star-bnl/star-sw/commit/ae1dae2f96d009378c6be9dcf21e19c0f523380d))

I reviewed #538 just now. It looks good to me. I suggest we merge it before this one

jdbrice committed:
FST tracking mode for high mult and data with missing sTGCs

jdbrice commented:
I finally updated this PR to include only the needed changes for this feature (FST based tracking)
Review can go ahead now

jdbrice committed:
Verify working refit (FST and FTT) in ideal sim, remove old unused track info

jdbrice commented:
commit e65362f made changes to verify everything works for the ideal simulation cases:
- FTT seed finding + FST refit
- FST seed finding + FTT refit

I added:
StRoot/StFwdTrackMaker/macro/sim/ideal_sim.C
StRoot/StFwdTrackMaker/macro/sim/gen

to make running the code in this way easy. It can be tested with:

```
cons
ln -s StRoot/StFwdTrackMaker/macro/sim/ .
./sim/gen 100
./sim/ideal_sim.C 100
```

jdbrice committed:
Logic update so that non-MC track finding (standard) works when TrackFinder node present

jdbrice committed:
remove StJPSiMaker dep from sim.C, keep feature branches separate

jdbrice committed:
add scripts for runnning common simulation configurations (ideal, realistic + fst/ftt seeds)

jdbrice committed:
Merge branch 'star-bnl:main' into fst-tracking-mode

jdbrice committed:
restore StBFChain doc files

jdbrice committed:
load hits from MuDst, allow cached geometry, and properly handle GenTree=false

jdbrice committed:
Changes to resolve severe memory leaks due to geometry navigtor, and Genfit tracks not deleted (use shared_ptr where possible)

jdbrice committed:
Update to StFwdTrack for MC track id and qa truth

jdbrice committed:
Update StMuFwdTrack to mirror StFwdTrack (Mc Id and QA)

jdbrice committed:
Update macros used for sim and MuDst analysis

jdbrice commented:
OK with [99752c4](https://github.com/star-bnl/star-sw/pull/571/commits/99752c467481f892c715181aa864b0a3d0ceeb08)
I think I have brought in all of the most important updates:
- Many changes throughout the StFwdTrackMaker package to improve memory management and remove old code / ensure debug and code meant for simulation only runs when requested.
- StEvent and StMuDst have minor updates to the St(Mu)FwdTrack classes to allow MC association information
- Updates to macros for standard use of the fwd tracking on simulation files, mudst etc.

genevb commented on StRoot/StMuDSTMaker/COMMON/StMuFwdTrack.cxx:
```diff
@@ -19,6 +19,9 @@ void StMuFwdTrack::set( StFwdTrack * evTrack) {
 
+    mIdTruth = evTrack->idTruth();
+    mQATruth = evTrack->qaTruth();
+    
```
From what I can tell, you rely on the default behavior of C++ to initialize the value of StEvent/StFwdTrack.cxx::mIdTruth to zero (for all real tracks), as it isn't explicitly initialized to have a default value neither there nor here. Is this the desired approach?

genevb commented on StRoot/StFwdTrackMaker/macro/mudst/mudst.C:
```diff
@@ -0,0 +1,129 @@
+	// needed since I use the StMuTrack
+	gSystem->Load("StarClassLibrary");
+	gSystem->Load("StStrangeMuDstMaker");
+	gSystem->Load("StMuDSTMaker");
```
`loadSharedLibraries()` already loads `StarClassLibrary`, `StStrangeMuDstMaker`, `StMuDSTMaker`, so not needed here. Not sure if I missed others, but this isn't critical.

genevb commented on StRoot/StFwdTrackMaker/macro/mudst/mudst.C:
```diff
@@ -0,0 +1,129 @@
+	cout << "LL1" << endl;
+	gROOT->LoadMacro("$STAR/StRoot/StMuDSTMaker/COMMON/macros/loadSharedLibraries.C");
+	loadSharedLibraries();
+	cout << "LL2" << endl;
```
These three LL0/1/2 cout lines can probably be dropped off now?

genevb reviewed (commented):
Just a few minor observations. I didn't dig through all the `sim` and `Tracker` changes, so I can't comment on those.

klendathu2k commented on StRoot/StMuDSTMaker/COMMON/StMuFwdTrack.h:
```diff
@@ -121,6 +121,8 @@ class StMuFwdTrack : public TObject {
+    UShort_t qaTruth() const { return mQATruth; }
 
 
     void setPrimaryMomentum( TVector3 mom ) { mPrimaryMomentum = mom; }
```
I suggest switching to StThreeVector for speed.  TVector3 has an additional overhead b/c it inherits from TObject.  Which means that every instance of every vector you create will be registered in a global ROOT map, and subsequently destroyed when you call the destructor.  

Looking at the MuDST codes... I see that we are using both TVector3 and StThreeVector<x>  (with x=float or double) to store vector quantities.... so consider this a soft suggestion.

klendathu2k commented on StRoot/StFwdTrackMaker/include/Tracker/TrackFitter.h:
```diff
@@ -769,11 +828,12 @@ class TrackFitter {
 
         // get the seed info from our hits
         TVector3 seedMom, seedPos;
```
[very small optimization...]

Since you are initializing these below... you can make these static variables and save the overhead of registering them with ROOT every time the ctor is called.

klendathu2k reviewed (commented):
Just took a quick look, and commented on some possible optimizations.  TVector3 is used several places and, where ever possible, you should avoid calling the constructor.  TVector3 registers itself with ROOT, which introduces an overhead.  (We ran into this slowing down Sti O(10%) in the past as I recall).  Otherwise I think this looks good.  I'll take a closer look tomorrow.

plexoos commented on StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx:
```diff
@@ -737,22 +742,113 @@ void StFwdTrackMaker::loadFttHitsFromGEANT( FwdDataSource::McTrackMap_t &mcTrack
+        // we use d+4 so that both FTT and FST start at 4
+        FwdHit *hit = new FwdHit(count++, x0, y0, vZ, d+4, 0, hitCov3, nullptr);
+        // Add the hit to the hit map
+        hitMap[d+4].push_back(hit);
```
I think this is the case where you want to be a bit more explicit with the code. You may know that you use smart pointers but it does look scary to a casual reviewer who is not aware of the internals.

```suggestion
        auto hit = std::make_unique<FwdHit>(count++, hitInfo.x, hitInfo.y, hitInfo.z, hitInfo.d + START_INDEX_FOR_FST, 0, hitCov3, nullptr);
        hitMap[d + START_INDEX_FOR_FST].push_back(std::move(hit));
```

And BTW, why not create the hits on the stack?

plexoos commented on StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx:
```diff
@@ -737,22 +742,113 @@ void StFwdTrackMaker::loadFttHitsFromGEANT( FwdDataSource::McTrackMap_t &mcTrack
+        const float dz1 = fabs( vZ - 165.248 );
+        const float dz2 = fabs( vZ - 178.781 );
+
+        int d = 0 * ( dz0 < 1.0 ) + 1 * ( dz1 < 1.0 ) + 2 * ( dz2 < 1.0 );
```
Magic numbers... It took me a while to understand the meaning of 1.0 which is probably a pretty arbitrary value.

plexoos commented on StRoot/StFwdTrackMaker/include/Tracker/FwdDataSource.h:
```diff
@@ -37,13 +37,15 @@ class FwdDataSource {
       for ( auto kv : mFttHits ){
           for ( auto h : kv.second ){
               delete h;
+              h = 0;
```
Oh, actually you are not using smart pointers for KiTrack hits... If you did

```cpp
using HitMap_t = std::map<int, std::vector<std::unique_ptr<KiTrack::IHit>>>;
```
it would simplify the cleanup code
```cpp
        for (auto& kv : mFttHits) {
            kv.second.clear();
        }
        for (auto& kv : mFstHits) {
            kv.second.clear();
        }
        mFttHits.clear();
        mFstHits.clear();
```

jdbrice replied:
Hits are now created on the stack and cleared each event by the StFwdTrackMaker

plexoos committed:
Merge branch 'main' into fst-tracking-mode

Commit:
use static const TVector3 where possible

Commit:
StMuFwdTrack initialize members

Commit:
Make FwdHits on stack, easier memory managment

Commit:
cleanup macros for fwd mudst processing

Commit:
Use FST z locations from geom to identify hit plane index

klendathu2k commented on StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx:
```diff
@@ -455,7 +458,17 @@ int StFwdTrackMaker::Init() {
-    mForwardTracker->initialize( geoCache, mGenHistograms );
+    mForwardTracker->initialize( mGeoCache, mGenHistograms );
+
+    // geometry should be available from here (mForwardTracker will initialize cache if needed)
```
I'm slightly concerned about pulling the z-locations from the geometry _and_ having fallback values hardcoded here.  

Is there at least logging in place to keep track of when / if we fall back to the default?

genevb reviewed (commented):
I posted 3 minor comments on this earlier. May not be critical to fix, but that's it for my part of the review, and a quick note that they'll be addressed now/later/not at all would be good.

plexoos commented:
Thank you, Daniel, for addressing the comments! Everything looks good as far as I can tell.

As a last request we discussed, could you point me to a recent data file with raw FST data from this run? I would like to extend our CI tests at https://github.com/star-bnl/star-sw/blob/main/tests/joblist.json#L1126, which already include data from 2023.

jdbrice committed:
Cleanup of the StFwdTrackMaker code
- Improve memory management of FwdHits
- Add FST and FTT detection methods to FwdHit class and update FwdGeomUtils
- cleanup track fitter logic
- Add additional track seed modes
- use enums where possible for clearer intentions
- remove histogram and ttree output from FwdTrackMaker and TrackFitter
- Tests against data and simulation

jdbrice committed:
Merge branch 'main' into fst-tracking-mode

jdbrice commented:
@plexoos @iraklic Please let me know if anything more is needed before we can merge this PR. There are others that depend on this. From what I can tell the 2 tests that failed were due to network issues - @plexoos maybe knows better.

iraklic commented:
> @plexoos @iraklic Please let me know if anything more is needed before we can merge this PR. There are others that depend on this. From what I can tell the 2 tests that failed were due to network issues - @plexoos maybe knows better.

Sorry for not answering so long. I am not ignoring your PR. I just have been super busy and this seems to be a substantial one, with the discussion that I have not really followed. I this it is better if I defer this to others, but I will go over this once I am back to the office if it is not done till then.

jdbrice commented:
@fgeurts @plexoos Please let me know if I need to do anything more. The recent commits were mostly clean-up (albeit a lot because I removed almost all debug histograms/ttree output). I am not sure why test 125 fails, I wonder if it is a separate issue? Either way I am looking into it.

plexoos commented:
The last time all checks passed was with this commit: @jdbrice
[Use FST z locations from geom to identify hit plane index](https://github.com/star-bnl/star-sw/pull/571/commits/0e9b6746b9023bee1fe6a3379035b0f3d377b616)
[0e9b674](https://github.com/star-bnl/star-sw/pull/571/commits/0e9b6746b9023bee1fe6a3379035b0f3d377b616)

However, the next commit failed: @jdbrice
[Cleanup of the StFwdTrackMaker code](https://github.com/star-bnl/star-sw/pull/571/commits/0241f63f080aafa93e79bb339539012443df9ba2)
[0241f63](https://github.com/star-bnl/star-sw/pull/571/commits/0241f63f080aafa93e79bb339539012443df9ba2)

Let's try to revert it to see if it really introduced a bug.

plexoos committed:
Revert "Cleanup of the StFwdTrackMaker code"

This reverts commit 0241f63f080aafa93e79bb339539012443df9ba2.

plexoos commented:
The test job 125 completed successfully after reverting the changes in @jdbrice's commit.

[Use FST z locations from geom to identify hit plane index](https://github.com/star-bnl/star-sw/pull/571/commits/0e9b6746b9023bee1fe6a3379035b0f3d377b616)
[0e9b674](https://github.com/star-bnl/star-sw/pull/571/commits/0e9b6746b9023bee1fe6a3379035b0f3d377b616)

This is the only test that uses the `FwdTrack` option:

```
bfc.C(200, "pp2022,StiCA,BEmcChkStat,epdhit,FwdTrack", "/star/rcf/test/daq/2023/072/23072022//st_physics_23072022_raw_4000001.daq")
```

From this, I conclude that something in that commit causes the test to fail.

jdbrice commented:
Thank you @plexoos  - I will investigate and try to resolve.

plexoos committed:
Merge branch 'main' into fst-tracking-mode

jdbrice committed:
Merge branch 'star-bnl:main' into fst-tracking-mode

jdbrice committed:
incorporate updates from dev

jdbrice committed:
Update macros and begin fixing primary vertex crash in StFwdTrackMaker

jdbrice committed:
nullptr check when searching for primary vertex options, improve default config handeling

jdbrice committed:
Merge branch 'main' into fst-tracking-mode

jdbrice committed:
fix incorrect merge, duplicate symbol

jdbrice committed:
merge in PR for projections to FCS with correct position and angle

jdbrice committed:
Utilities for debugging (print statements and track seed bailout option)

jdbrice committed:
bailout of track fits with bad setup (collinear zero Q) and skip track projections with zero points with measurements

jdbrice committed:
Load config in ctor to allow user to override defaults in driving macro. add LOG_INFO to identify test 125 failure

jdbrice committed:
remove verbose debug print code

jdbrice commented:
@plexoos, @klendathu2k  - I have finally identified the seg fault, it occurs when I try to find from various possible sources an event primary vertex. This is the code that fails:
```cpp
    if(mMuDstMaker && mMuDstMaker->muDst() && mMuDstMaker->muDst()->primaryVertex() ) {
        mEventVertex.SetX(mMuDstMaker->muDst()->primaryVertex()->position().x());
        mEventVertex.SetY(mMuDstMaker->muDst()->primaryVertex()->position().y());
        mEventVertex.SetZ(mMuDstMaker->muDst()->primaryVertex()->position().z());
        mFwdVertexSource = kFwdVertexSourceTpc;
        return mEventVertex;
    } 
```

as you can see I am testing for nullptrs, and in the most recent code I print out each pointer:
```
StFwdTrackMaker:INFO  - Searching for Event Vertex from MuDstMaker: 0xc279e60
StFwdTrackMaker:INFO  - Searching for Event Vertex from MuDst: 0xc2825a0
StFwdTrackMaker:INFO  - Searching for Event Vertex from MuDst Primary Vertex: 0x700000000000093
StFwdTrackMaker:FATAL - TUnixSystem::DispatchSignals : segmentation violation
 Generating stack trace...
```

The `mMuDstMaker->muDst()->primaryVertex()` is NOT null but also not valid, so when it gets dereferenced it causes a set fault.  However, in every other environment I can run (docker any container, RCF etc.) for the exact same chain (test 125) I get a nullptr for the `mMuDstMaker->muDst()->primaryVertex()` and the code runs fine. 

So I will temporarily remove this code to fix this PR so we can move forward with the next library, but this is likely indicative of some deeper issues with the MuDst Primary Vertex. Let me know if you have any thoughts

plexoos commented:
Good catch, @jdbrice! Yeah, dereferencing an invalid pointer is a classic case of undefined behavior in C++, where it might work on one machine but crash on another.

jdbrice committed:
avoid MuDst primaryVertex bad pointer - check # of vertices instead of pointer being null

jdbrice commented:
@plexoos - thanks for the suggestions that helped me find it. 

OK I 'd like to close this PR without adding anything else to it (except anything identified necessary by further review). I prefer to make any final changes needed for a new library in smaller more scoped PRs. So if you and @klendathu2k or @genevb etc can take a look, I will leave this alone for now. But it is ready from my side

klendathu2k commented on StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx:
```diff
@@ -1068,42 +967,144 @@ void StFwdTrackMaker::loadFcs( ) {
-                mTreeData.fcsY.push_back( y0 );
-                mTreeData.fcsZ.push_back( zepd );
-                mTreeData.fcsDet.push_back( det );
+TVector3 StFwdTrackMaker::GetEventPrimaryVertex(){
```
I haven't traced this through, but... if GetEventPrimaryVertex is used in the track fit, then I have some concerns with this implementation.

Here is my reading of what this does...

If the MC vertex is available, return it.  If not, use the MuDst.  If not... check to see if StEvent is available and return what we have.  If not, return some default values.

Here are my concerns...

1) StEvent is our event data model.  We do not need to create the MuDst to run our code, but StEvent must be available.  It is the preferred source for the primary vertex.

2) Simulated vertices are returned if they are found, taking precedence if the vertex is available in StEvent / MuDst.  This means that you will be treating simulated data and real data differently... you'll take the exact primary vertex for simulation (which is always the correct one) and the reconstructed primary vertex from data (which might not point to the real primary vertex).

jdbrice replied:
@klendathu2k Thanks for pointing this out - I agree with your concerns. I will address this in new PR - The primary vertex handling can still be improved in a couple of ways in addition to your comment here. I will follow this up promptly with a (draft) PR so that this is addressed.

klendathu2k commented on StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx:
```diff
@@ -1137,264 +1138,263 @@ int StFwdTrackMaker::Make() {
-    FillTrackDeltas();
 
-    LOG_INFO << "Forward tracking on this event took " << (FwdTrackerUtils::nowNanoSecond() - itStart) * 1e-6 << " ms" << endm;
+    /**********************************************************************/
```
Not a show stopper ... but I would split visualization into a separate maker.

jdbrice replied:
@klendathu2k - thanks for the suggestion. While I was able to separate almost all of the other non-essential parts of the tracking (matching, QA+ttree generation, etc.) I have not yet been able to move this out of the core FwdTrackMaker because it needs the in-memory GenFit track object. I will continue to look into ways to separate it out.

The visualization is OFF by default (only turned on if the user turns it on)

jdbrice committed:
Merge branch 'main' into fst-tracking-mode

jdbrice committed:
fix merge duplicate defs

klendathu2k reviewed (approved):
I remain concerned about the GetPrimaryEventVertex method.  Specifically in a full simulation, we should be fitting to the reconstructed rather than the simulated event vertex.  That said, we should not hold up getting this ready for data production.

jdbrice commented:
@plexoos  @genevb @iraklic @fisyak @R-Witt Can you please review the code / comments and resolutions. If things are good please approve - we'd like to get this merged
